### PR TITLE
Make ``Message`` a ``dataclass``

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

### CODE REVIEW

**Review by Pierre-Sassoulas — APPROVED** [review]
👌  (This is a lot faster than creating a class using an 'exec', with old style named tuple too )

File: pylint/message/message.py [hunk 1]
```diff
@@ -4,90 +4,78 @@
...
+                None,
+            )
+
+        self.msg_id = msg_id
+        self.symbol = symbol
+        self.msg = msg
+        self.C = msg_id[0]
+        self.category = MSG_TYPES[msg_id[0]]
+        self.confidence = confidence or UNDEFINED
+        self.abspath = location.abspath
```
Comment (Pierre-Sassoulas) [inline pylint/message/message.py:65]: We could use the location as an attribute and create property to not break the API.

Comment (DanielNoord) [reply pylint/message/message.py:65]: Not sure what you mean, see L73. I added location to be able to access it directly and with the new `if` we can now assume it is `MessageLocationTuple` here.

Comment (Pierre-Sassoulas) [reply pylint/message/message.py:65]: Doesn't it mean that we could modify "line" without modifying "location.line" ? Maybe we need to use a frozen dataclass to be sure it won't be incoherent.

Comment (DanielNoord) [reply pylint/message/message.py:65]: Ah, yeah I didn't consider that!

I'll remove the `location` attribute. We don't even need it ourselves. That makes it easier. If we want to expand the API later we can think about it then.

File: pylint/message/message.py [hunk 2]
```diff
@@ -4,90 +4,77 @@
...
+    confidence: Confidence
+    abspath: str
+    path: str
+    module: str
+    obj: str
+    line: int
+    column: int
+    end_line: int | None
+    end_column: int | None
+    location: MessageLocationTuple
```
Comment (DanielNoord) [inline pylint/message/message.py:33]:
```suggestion
```

**Review by Pierre-Sassoulas — APPROVED** [review]
Great ! I kinda want to know how much faster pylint's run right now but I'm too lazy to do it 😄 If I had to guess, something like 0.10% to 0.20%

**Comment by DanielNoord** [comment]
> Great ! I kinda want to know how much faster pylint's run right now but I'm too lazy to do it 😄 If I had to guess, something like 0.10% to 0.20%

Anyway, my own contributing is 10% faster by not having to think about the horrible construct of `_MsgBase`... 😄

**Pull request merged**